### ISSUE TITLE
Cannot open new PO when changing pattern

### ISSUE THREAD

**Issue by derekwolfson** [inventree/InvenTree#4286]
### Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find a similar issue

### Describe the bug*

I had used ``PO-{date}-{ref:04d}`` -- not realizing this was going to give the entire datetime.

I made a PO with this -- which I later renamed to ``PO-0007``

I then changed it to what I wanted it to be -- which is ``PO-{ref:05d}{date:%y}{date:%m}``

And when I try to create a new PO it populates correctly as -- ``PO-000082301`` but when I go to submit the PO I get the following error: 

``redefinition of group name 'date' as group 3; was group 2 at position 33``

It seems the backend is choking on changing the pattern when including multiple ``{date}`` calls, even if they aren't identical (like ``{date}`` vs ``{date:%y} {date:%m}``).  

I suppose this could generalize to any django call that can have sub-elements triggered by ``%`` but have not tested as such.

### Steps to Reproduce

I was able to recreate the issue on my Docker server and the Inventree Demo server with Admin access:

3. Set PO reference with a ``{date}`` element (i.e. in position 2 -- ``PO-{ref:04d}{date}``)
4. Open a PO
5. submit PO
6. Now set PO Reference with TWO ``{date}`` elements -- like ``PO-{ref:04d}{date}{date}``)
7. Kick dirt...

### Expected behaviour

Changing from
``PO-{date}{ref:04d}``
to 
``PO-{ref:04d}{date:%y}{date:%m}``

Should be allowed and the PO pattern validation engine should treat these sub-elements as distinct.

### Deployment Method

- [X] Docker
- [ ] Bare metal

### Version Information

# Version Information:
InvenTree-Version: 0.9.2
Django Version: 3.2.16
Commit Hash: 236ca36
Commit Date: 2023-01-04
Database: postgresql
Debug-Mode: False
Deployed using Docker: True
Active plugins: False

### Relevant log output

```shell
Error object (21)
HISTORY

Type:
error
Path:
/api/order/po/
Info:
redefinition of group name 'date' as group 3; was group 2 at position 33
When:
Jan. 31, 2023, 9:56 p.m.
Data:
Traceback (most recent call last):

File "/root/.local/lib/python3.9/site-packages/rest_framework/views.py", line 506, in dispatch
response = handler(request, *args, **kwargs)

File "/root/.local/lib/python3.9/site-packages/rest_framework/generics.py", line 242, in post
return self.create(request, *args, **kwargs)

File "/home/inventree/InvenTree/order/api.py", line 138, in create
serializer.is_valid(raise_exception=True)

File "/root/.local/lib/python3.9/site-packages/rest_framework/serializers.py", line 227, in is_valid
self._validated_data = self.run_validation(self.initial_data)

File "/home/inventree/InvenTree/InvenTree/serializers.py", line 175, in run_validation
data = super().run_validation(data)

File "/root/.local/lib/python3.9/site-packages/rest_framework/serializers.py", line 426, in run_validation
value = self.to_internal_value(data)

File "/root/.local/lib/python3.9/site-packages/rest_framework/serializers.py", line 485, in to_internal_value
validated_value = validate_method(validated_value)

File "/home/inventree/InvenTree/order/serializers.py", line 131, in validate_reference
order.models.PurchaseOrder.validate_reference_field(reference)

File "/home/inventree/InvenTree/InvenTree/models.py", line 269, in validate_reference_field
if not InvenTree.format.validate_string(value, pattern):

File "/home/inventree/InvenTree/InvenTree/format.py", line 117, in validate_string
result = re.match(pattern, value)

File "/usr/local/lib/python3.9/re.py", line 191, in match
return _compile(pattern, flags).match(string)

File "/usr/local/lib/python3.9/re.py", line 304, in _compile
p = sre_compile.compile(pattern, flags)

File "/usr/local/lib/python3.9/sre_compile.py", line 788, in compile
p = sre_parse.parse(p, flags)

File "/usr/local/lib/python3.9/sre_parse.py", line 955, in parse
p = _parse_sub(source, state, flags & SRE_FLAG_VERBOSE, 0)

File "/usr/local/lib/python3.9/sre_parse.py", line 444, in _parse_sub
itemsappend(_parse(source, state, verbose, nested + 1,

File "/usr/local/lib/python3.9/sre_parse.py", line 838, in _parse
raise source.error(err.msg, len(name) + 1) from None

re.error: redefinition of group name 'date' as group 3; was group 2 at position 33
```

**Comment by derekwolfson** [comment]
###ALSO -- THANK YOU :) ###

**Comment by derekwolfson** [comment]
Also worth nothing -- it is broken for other types of orders now too..

I tried to switch my build orders from ``BO-{ref:04d}`` to  ``BO-{ref:04d}{date:%m}{date:%y}``

and received the same error in the console.

**Comment by SchrodingersGat** [comment]
@derekwolfson thanks for reporting this, I must admit I had not considered such an edge case. Due to the way this is implemented, doubling up on named variables will not work!

However there is a way around this - you can format the date string in a more complex manner if you wish (see test below):

```python
from datetime import datetime

date = datetime.now()

ref = 123

reference = "PO-{ref:04d}-{date:%y%m%d}".format(**{
    'ref': ref,
    'date': date,
})

print("Reference:", reference)
```
```
> Reference: PO-0123-230202
```

So, you could could extract complex date information with a single date reference string.

*However, this would only work as long as your date is not split by some other text (you can still only use this once)*

**Comment by SchrodingersGat** [comment]
A list of available datetime codes is available here : https://www.programiz.com/python-programming/datetime/strftime

**Comment by SchrodingersGat** [comment]
@derekwolfson I've also updated the docs here - https://docs.inventree.org/en/latest/settings/reference/

**Comment by derekwolfson** [comment]
Fabulous.  I’m an absolute noob with Django and Python — so I really appreciate this update!!I had a feeling I could do it with some sort of substring style command but I thought I’d wait for your response — so thank you again. On Feb 1, 2023, at 2:51 PM, Oliver ***@***.***> wrote:﻿
@derekwolfson I've also updated the docs here - https://docs.inventree.org/en/latest/settings/reference/

**Comment by derekwolfson** [comment]
> @derekwolfson I've also updated the docs here - https://docs.inventree.org/en/latest/settings/reference/
> 
> _UPDATE_:
> 
> And, submitted a fix for the error itself: #4291

@schrodingersGat -- small typo in the documentation -- just need to change %M to %m (lowercase).

Right now it's pulling the %Y (year) than %M (minute).  

Small issue -- but thought I'd flag it while I see it.
